File: 241115/루돌프의 반란/rudolph-rebellion.java
import java.util.*;
import java.io.*;

public class Main {
    static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    static StringTokenizer st;
    static int n, m, p, c, d;
    static int ry, rx;
    static int [][] map;  // 디버깅
    static Santa[] santas;
    static boolean[] alive;
    static int[] stun;
    static int[] scores;

    static class Santa {
        int y, x, num;
        double distance;

        public Santa(int y, int x) {
            this.y = y;
            this.x = x;
        }

        public Santa(int y, int x, int num) {
            this.y = y;
            this.x = x;
            this.num = num;
        }

        public Santa(int y, int x, int num, double distance) {
            this.y = y;
            this.x = x;
            this.num = num;
            this.distance = distance;
        }
    }

    public static void main(String[] args) throws IOException {
        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());
        p = Integer.parseInt(st.nextToken());
        c = Integer.parseInt(st.nextToken());
        d = Integer.parseInt(st.nextToken());
        st = new StringTokenizer(br.readLine());
        ry = Integer.parseInt(st.nextToken());
        rx = Integer.parseInt(st.nextToken());
        santas = new Santa[p + 1];  // sy, sx
        for(int i = 1; i <= p; i++) {
            st = new StringTokenizer(br.readLine());
            int sNum, sy, sx;
            sNum = Integer.parseInt(st.nextToken());
            sy = Integer.parseInt(st.nextToken());
            sx = Integer.parseInt(st.nextToken());
            santas[sNum] = new Santa(sy, sx);
        }

        map = new int[n + 2][n + 2];
        alive = new boolean[p + 1];
        Arrays.fill(alive, true);
        stun = new int[p + 1];
        scores = new int[p + 1];

        solve();
    }

    static void initMap() {  //  디버깅, 산타 및 루돌프 위치 표기
        for(int i = 1; i <= p; i++) {
            Santa s = santas[i];
            map[s.y][s.x] = i;
        }
        map[ry][rx] = -1;
    }

    static void moveRudolph(int turn) {
        Santa santa = findSanta();
        int dy = 0, dx = 0;

        if(ry < santa.y)
            dy = 1;
        else if(ry > santa.y)
            dy = -1;

        if(rx < santa.x)
            dx = 1;
        else if(rx > santa.x)
            dx = -1;


        map[ry][rx] = 0;
        map[ry + dy][rx + dx] = -1;
        ry += dy;
        rx += dx;
        if(ry == santa.y && rx == santa.x) {
            pushSanta(santa, dy, dx, c);
            scores[santa.num] += c;
            stun[santa.num] = turn + 2;
        }
    }

    static Santa findSanta() {
        ArrayList<Santa> distances = new ArrayList<>();
        double minDistance = 2 * n * n;

        for(int i = 1; i <= p; i++) {
            if(!alive[i])
                continue;
            double distance = Math.pow((ry - santas[i].y), 2) + Math.pow((rx - santas[i].x), 2);
            if(distance <= minDistance) { // 동일한 거리가 있을 수도 있음
                minDistance = distance;
                distances.add(new Santa(santas[i].y, santas[i].x, i, distance));
            }
        }

        Collections.sort(distances, 
        Comparator.comparingDouble((Santa o) -> o.distance)
        .thenComparingInt((o) -> -o.y)
        .thenComparingInt((o) -> -o.x));
        return distances.get(0);
    }

    static void moveSanta(int turn) {
        int[][] directions = {{-1, 0}, {0, 1}, {1, 0}, {0, -1}};

        for(int i = 1; i <= p; i++) {
            if(!alive[i] || stun[i] > turn)
                continue;
            double minDistance = Math.pow((ry - santas[i].y), 2) + Math.pow((rx - santas[i].x), 2);
            int santa = 0, my = 0, mx = 0, dy = 0, dx = 0;
            for(int[] d: directions) {
                my = santas[i].y + d[0];
                mx = santas[i].x + d[1];
                double distance = Math.pow((ry - my), 2) + Math.pow((rx - mx), 2);
                if(1 <= my && my <= n && 1 <= mx && mx <= n && map[my][mx] <= 0 && distance < minDistance) {
                    minDistance = distance;
                    santa = i;
                    dy = d[0];
                    dx = d[1];
                }
            }

            if(santa != 0) {
                my = santas[santa].y + dy;
                mx = santas[santa].x + dx;
                map[santas[santa].y][santas[santa].x] = 0;
                if(map[my][mx] == 0) {
                    map[my][mx] = santa;
                    santas[santa] = new Santa(my, mx);
                } else {
                    pushSanta(new Santa(my, mx, santa), -dy, -dx, d);
                    scores[santa] += d;
                    stun[santa] = turn + 2;
                }
            }
        }
    }

    static void pushSanta(Santa santa, int dy, int dx, int power) {
        ArrayDeque<Santa> dq = new ArrayDeque<>();
        dq.add(santa);

        while(!dq.isEmpty()) {
            Santa current = dq.pollFirst();
            int my = current.y + (dy * power), mx = current.x + (dx * power);
            if(1 <= my && my <= n && 1 <= mx && mx <= n) {
                if(map[my][mx] != 0) {  // 다른 산타가 있는 경우
                	power = 1;  // 한 칸씩 밀어야 함
                    dq.add(new Santa(my, mx, map[my][mx]));
                }
	            map[my][mx] = current.num;
	            santas[current.num] = new Santa(my, mx);  //  산타 위치 정보 갱신
            }
            else {
                alive[current.num] = false;
                return;
            }
        }
    }

    static void solve() {
        initMap();

        for(int turn = 1; turn <= m; turn++) {
            int aliveSanta = p;
            Game:
            for(int i = 1; i <= p; i++) {
                aliveSanta -= alive[i] ? 1 : 0;
                if(aliveSanta == 0)
                    break Game;
            }
            moveRudolph(turn);
            moveSanta(turn);
            for(int i = 1; i <= p; i++) {
                if(alive[i] == true)
                	scores[i] += 1;
            }
        }
        for(int i = 1; i <= p; i++)
        	System.out.print(scores[i] + " ");
    }
}

File: 241115/루돌프의 반란/rudolph-rebellion.py
from collections import deque

class Main:
    def __init__(self):
        self.n, self.m, self.p, self.c, self.d = map(int, input().split())
        self.y, self.x = map(int, input().split())
        self.santas = [[0, 0] for _ in range(self.p + 1)]
        for _ in range(1, self.p + 1):
            santa, sy, sx = list(map(int, input().split()))
            self.santas[santa][0], self.santas[santa][1] = sy, sx
        self.board = [[0] * (self.n + 2) for _ in range(self.n + 2)]
        self.scores = [0] * (self.p + 1)  # 점수
        self.alive = [True] * (self.p + 1)  # 생존 여부
        self.stun = [0] * (self.p + 1)  # 기절 시간

    @staticmethod
    def cal_distance(r1, c1, r2, c2):
        return (r1 - r2)**2 + (c1 - c2)**2

    def find_santa(self):
        distances = []
        min_distance = 2 * self.n**2
        for santa in range(1, self.p + 1):  # 가장 가까운 산타 찾기
            if not self.alive[santa]:
                continue
            sy, sx = self.santas[santa]
            distance = self.cal_distance(self.y, self.x, sy, sx)
            if distance <= min_distance:
                min_distance = distance
                distances.append((min_distance, sy, sx, santa))
        distances.sort(key=lambda x: (x[0], -x[1], -x[2]))
        return distances[0]

    def move_rudolph(self, turn):
        closest_santa = self.find_santa()
        _, sy, sx, santa = closest_santa

        dy, dx = 0, 0
        if self.y < sy:
            dy = 1
        elif self.y > sy:
            dy = -1

        if self.x < sx:
            dx = 1
        elif self.x > sx:
            dx = -1

        self.board[self.y][self.x] = 0
        self.y, self.x = self.y + dy, self.x + dx
        self.board[self.y][self.x] = -1

        if self.y == sy and self.x == sx:  # 루돌프가 산타와 충돌
            self.back_santa(santa, sy, sx, dy, dx, self.c)  # 산타 밀기
            self.stun[santa] = turn + 2
            self.scores[santa] += self.c

    def back_santa(self, santa, sy, sx, dy, dx, toward):
        dq = deque([(santa, sy, sx, toward)])

        while dq:
            santa, sy, sx, toward = dq.popleft()
            my, mx = sy + (dy * toward), sx + (dx * toward)
            if 1 <= my <= self.n and 1 <= mx <= self.n:
                if self.board[my][mx] != 0:  # 빈칸이면
                    dq.append((self.board[my][mx], my, mx, 1))
                self.board[my][mx] = santa
                self.santas[santa] = [my, mx]
            else:  # 보드를 벗어나면
                self.alive[santa] = False
                return

    def move_santa(self, turn, santa):
        sy, sx = self.santas[santa]
        distances = []
        min_distance = self.cal_distance(self.y, self.x, sy, sx)

        for dy, dx in ((-1, 0), (0, 1), (1, 0), (0, -1)):
            my, mx = sy + dy, sx + dx
            distance = self.cal_distance(self.y, self.x, my, mx)
            if 1 <= my <= self.n and 1 <= mx <= self.n and self.board[my][mx] <= 0 and distance < min_distance:
                min_distance = distance
                distances.append((my, mx, dy, dx))

        if len(distances) != 0:
            my, mx, dy, dx = distances[-1]  # 마지막으로 저장된 좌표가 가장 가까운 좌표(조건문상 계속 최소 거리일 때 활성화되므로)
            if my == self.y and mx == self.x:
                self.board[sy][sx] = 0
                self.back_santa(santa, my, mx, -dy, -dx, self.d)
                self.stun[santa] = turn + 2
                self.scores[santa] += self.d
            else:
                self.board[sy][sx] = 0
                self.board[my][mx] = santa
                self.santas[santa] = [my, mx]

    def init_board(self):
        self.alive[0] = False
        for santa in range(1, self.p + 1):  # 산타 위치
            sy, sx = self.santas[santa]
            self.board[sy][sx] = santa

        self.board[self.y][self.x] = -1

    def solve(self):
        self.init_board()
        # for i in self.board:
        #     print(i)
        # print()
        for turn in range(1, self.m + 1):
            if self.alive.count(True) == 0:
                break
            self.move_rudolph(turn)
            for santa in range(1, self.p + 1):
                if not self.alive[santa]:
                    continue
                if self.stun[santa] > turn:
                    continue
                self.move_santa(turn, santa)

            for santa in range(1, self.p + 1):
                if self.alive[santa]:
                    self.scores[santa] += 1

            # print(turn)
            # for i in self.board:  # 디버깅 코드
            #     print(i)
            # print()

        print(*self.scores[1:], end=' ')

problem = Main()
problem.solve()